### PR TITLE
Review relay v4 changes

### PR DESCRIPTION
**Pull request by alloy**
These were some changes I needed to [update our codebase](https://github.com/artsy/emission/pull/1673).

The individual commits state what they are about, let me know if there’s anything that’s not clear.

### CODE REVIEW

**Comment by alloy** [comment]
The CI failure appears to be unrelated to the actual changes.

File: types/react-relay/test/react-relay-tests.tsx [hunk 1]
```diff
@@ -170,12 +171,58 @@ const Story = (() => {
...
+        const relayProp: RelayRefetchProp = {} as any;
+        // $ExpectError
+        <StoryRefetchContainer story={story} onLike={onLike} relay={relayProp} />;
+    }
+
+    function requiresTheRelayPropInPropsInterface() {
+        // FIXME: This should throw a type error, but doesn't
+        // const FunctionComponent: React.FC<{}> = () => null;
+        // // $ExpectError
+        // createRefetchContainer(FunctionComponent, { story: graphql`` }, graphql``);
```
Comment (alloy) [inline types/react-relay/test/react-relay-tests.tsx:199]: Sooo… really not sure why this doesn’t error. It _should_, but it doesn’t 🤷‍♂

File: types/react-relay/index.d.ts [hunk 2]
```diff
@@ -128,10 +128,12 @@ interface GeneratedNodeMap {
 
 export type ContainerProps<Props> = MappedFragmentProps<Pick<Props, Exclude<keyof Props, 'relay'>>>;
 
-export function createFragmentContainer<Props extends { relay: RelayProp }>(
-    Component: React.ComponentType<Props>,
+export type Container<Props> = React.ComponentType<ContainerProps<Props> & { componentRef?: (ref: any) => void }>;
```
Comment (ckknight) [inline types/react-relay/index.d.ts:131]: Ugh, Relay doesn't use `forwardRef` yet, then? I believe it'll also work where `componentRef` is a `React.RefObject<any>`

File: types/react-relay/index.d.ts [hunk 3]
```diff
@@ -30,16 +26,19 @@ import {
 // ./ReactRelayTypes
 export interface RelayProp {
     environment: Environment;
+    refetch: undefined; // ensures no RelayRefetchProp is used with a fragment container
+    hasMore: undefined; // ensures no RelayPaginationProp is used with a fragment container
```
Comment (ckknight) [inline types/react-relay/index.d.ts:30]: Is this more appropriate than

```suggestion
    hasMore?: never; // ensures no RelayPaginationProp is used with a fragment container
```

Also, it seems like these should be optional regardless.

**Review by ckknight — APPROVED** [review]

File: types/react-relay/index.d.ts (hunk 3 above)
Comment (alloy) [reply types/react-relay/index.d.ts:30]: Pretty sure I tried that and many other permutations. The tests would only pass with non-optional `undefined`.

File: types/react-relay/index.d.ts (hunk 2 above)
Comment (alloy) [reply types/react-relay/index.d.ts:131]: I think it _also_ does nowadays, so we should probably look into that and update, but they definitely still support this API and thus so should we. It’s unfortunate that this isn’t reflected in any types, just in docs and implementation.

**Pull request merged**